rfd/0205-improved-onprem-joining.md

```diff
@@ -0,0 +1,1060 @@
+
+## Required Approvers
+
+- Engineering: @strideynet && @zmb3
+
+- Product: @thedevelopnik
+
+## What
+
+This RFD proposes several improvements to better support non-delegated and
+on-prem joining, particularly for Machine ID.
+
+Primarily, we discuss a new `bound-keypair` join method intended to replace the
+traditional `token` join method for many use cases, but also proposes a number
+of related UX improvements to improve bot joining generally and `token` and
+`bound-keypair` joining in particular.
+
+## Why
+
+Today, if some form of delegated joining is not available, bots must fall back
+to the traditional `token` join method. This join method simple and universal:
+it has effectively zero hardware or software requirements, works with any (or
+no) cloud provider, and is perfect for demos and experimentation. It's also
+relatively secure: single use tokens ensure a Teleport admin user is directly
+involved with each bot join, and generation counter checks help ensure bot
+identities are difficult to exfiltrate unnoticed.
+
+Unfortunately, that's a fairly exhaustive list of positives, and when used in a
+production environment, bot token joining has major operational problems:
+
+- Onboarding scales poorly: joining a large fleet of bot instances means
+  provisioning a token for each bot, which also means generating secrets
+  yourself, and distributing them appropriately.
+
+- Ongoing maintenance requires manual intervention, breaking IaC principles. We
+  should assume `token` joined bots will inevitably fail at some point, and when
+  this happens, a new token must be issued - manually.
+
+- Internal bot identities have a hard 24 hour TTL limit, limiting the maximum
+  possible resiliency to 24 hours before a bot can no longer rejoin without
+  manual human intervention
+
+- `token`-method tokens are themselves secret values and their names need to be
+  treated carefully to avoid leaking secrets
+
+- Bots occasionally trigger generation counter lockouts, killing themselves and
+  any instances on the same bot
+
+- When a bot instance's renewable identity expires, there's no clear way to tell
+  that it stopped functioning other than checking the `tbot` process directly.
+
+These limitations led to a surprisingly narrow set of use cases where token
+joining was really a *good* experience, effectively just:
+
+- Experimentation and development use. It's simple and comprehensible which
+  makes it great for use in documentation - `tctl bots add` even gives you the
+  command to run to start a bot immediately!
+
+- Running very few, very reliable, long-lived systems. If you can reasonably
+  expect your system to never go down for more than 24 hours, bots can happily
+  run for months.
+
+  (Ironically, Kubernetes is a great environment in which to run `token`-joined
+  bots since it'll rapidly reschedule any bot deployments that fail... but we
+  have a dedicated `kubernetes` delegated join method.)
```

Comment:
I disagree, Kubernetes is not a viable environment for token joining. This misconception caused us to build the kube secret state persistence and do irremediable harm to the teleport-kube-agent by making it stateful.

I think this claim can be removed without altering the RFD's essence. I don't want to bikeshed on topics which are not central to the RFD, feel free to Slack me if you want to exchange about token joining challenges in kube.

Reply:
Heh, that's a fair call out, I've removed that comment.

That said, I'd definitely like to hear more about the issues you ran into. Statelessness is definitely the ideal, especially for your use case, but I'd like this to be as viable as possible. The hope is that this new join method would have made your experience at least _less_ miserable.